tests/main/views/test_audits.py

```diff
@@ -50,8 +54,204 @@ def add_audit_events_with_db_object(self):
             db.session.commit()
             return tuple(event.id for event in events)
 
+    def add_audit_events_by_param_tuples(self, service_audit_event_params, supplier_audit_event_params):
+        with self.app.app_context():
+            # some migrations create audit events, but we want to start with a clean slate
+            AuditEvent.query.delete()
+            service_ids = db.session.query(Service.id).order_by(Service.id).all()
+            supplier_ids = db.session.query(Supplier.id).order_by(Supplier.id).all()
+
+            audit_events = []
+
+            for (ref_model, ref_model_ids), (obj_id, audit_type, created_at, acknowledged_at) in chain(
+                    izip(repeat((Service, service_ids,)), service_audit_event_params),
+                    izip(repeat((Supplier, supplier_ids,)), supplier_audit_event_params),
+                    ):
+                ae = AuditEvent(audit_type, "henry.flower@example.com", {}, ref_model(id=ref_model_ids[obj_id]))
+                ae.created_at = created_at
+                ae.acknowledged_at = acknowledged_at
+                ae.acknowledged = bool(acknowledged_at)
+                db.session.add(ae)
+                audit_events.append(ae)
+
+            db.session.commit()
+            # make a note of the ids that were given to these events, or rather the order they were generated
+            audit_event_id_lookup = {ae.id: i for i, ae in enumerate(audit_events)}
+            assert AuditEvent.query.count() == len(service_audit_event_params)+len(supplier_audit_event_params)
+
+            return audit_event_id_lookup
+
 
 class TestAuditEvents(BaseTestAuditEvents):
+    @pytest.mark.parametrize(
+        "service_audit_event_params,supplier_audit_event_params,req_params,expected_resp_events",
+        # where we refer to "id"s in the expected_response_params, because we can't be too sure about the *actual* ids
+        # given to objects, we're using 0-based notional "ids" mased on the order the audit events were inserted into
+        # the db. service_audit_event_params events are inserted in the order given, followed by the
+        # supplier_audit_event_params events. so if we had 5 service_audit_event_params and 2
+        # supplier_audit_event_params, "5" would refer to the audit event created by the first-listed
+        # supplier_audit_event_params.
+        # similarly, where supplier and service "id"s are referred to, the "id"s we're referring to are normalized
+        # pseudo-ids from 0-4 inclusive
+        chain.from_iterable(
+            ((serv_aeps, supp_aeps, req_params, expected_resp_events) for req_params, expected_resp_events in req_cases)
+            for serv_aeps, supp_aeps, req_cases in (
+                (
+                    (  # service_audit_event_params
```

Comment:
This would probably be easier to understand if it listed what the parameters are. Or maybe even better, if it used dictionaries instead of tuples.